modules/core/src/main/scala/fs2/kafka/internal/KafkaConsumerActor.scala

```diff
@@ -169,69 +160,52 @@ private[kafka] final class KafkaConsumerActor[F[_], K, V](
             .updateAndGet(_.asSubscribed)
             .log(ManuallyAssignedPartitions(partitions, _))
       }
-      .flatMap(deferred.complete)
+      .flatMap(callback)
   }
 
   private[this] def fetch(
     partition: TopicPartition,
     streamId: StreamId,
     partitionStreamId: PartitionStreamId,
-    deferred: Deferred[F, (Chunk[CommittableConsumerRecord[F, K, V]], FetchCompletedReason)]
+    callback: ((Chunk[CommittableConsumerRecord[F, K, V]], FetchCompletedReason)) => F[Unit]
   ): F[Unit] = {
     val assigned =
-      withConsumer { consumer =>
-        F.delay(consumer.assignment.contains(partition))
-      }
+      withConsumer.blocking { _.assignment.contains(partition) }
 
     def storeFetch =
       ref
         .modify { state =>
           val (newState, oldFetch) =
-            state.withFetch(partition, streamId, partitionStreamId, deferred)
+            state.withFetch(partition, streamId, partitionStreamId, callback)
           (newState, (newState, oldFetch))
         }
         .flatMap {
           case (newState, oldFetches) =>
-            log(StoredFetch(partition, deferred, newState)) >>
+            log(StoredFetch(partition, callback, newState)) >>
               oldFetches.traverse_ { fetch =>
                 fetch.completeRevoked(Chunk.empty) >>
                   log(RevokedPreviousFetch(partition, streamId))
               }
         }
 
     def completeRevoked =
-      deferred.complete((Chunk.empty, FetchCompletedReason.TopicPartitionRevoked))
+      callback((Chunk.empty, FetchCompletedReason.TopicPartitionRevoked))
 
     assigned.ifM(storeFetch, completeRevoked)
   }
 
-  private[this] def commitAsync(request: Request.Commit[F, K, V]): F[Unit] = {
-    val commit =
-      withConsumer { consumer =>
-        F.delay {
+  private[this] def commitAsync(request: Request.Commit[F, K, V]): F[Unit] =
+    withConsumer { (consumer, _) =>
+      F.delay {
           consumer.commitAsync(
             request.offsets.asJava,
-            new OffsetCommitCallback {
-              override def onComplete(
-                offsets: util.Map[TopicPartition, OffsetAndMetadata],
-                exception: Exception
-              ): Unit = {
-                val result = Option(exception).toLeft(())
-                val complete = request.deferred.complete(result)
-                F.runAsync(complete)(_ => IO.unit).unsafeRunSync()
-              }
-            }
+            (_, exception) => request.callback(Option(exception).toLeft(()))
           )
-        }.attempt
-      }
-
-    commit.flatMap {
-      case Right(()) => F.unit
-      case Left(e)   => request.deferred.complete(Left(e))
+        }
+        .handleErrorWith(e => F.delay(request.callback(Left(e))))
     }
```

Comment:
As above - am I correct in changing this to not shift anything to the blocking pool? (and generally, could you check the way I've replaced the `Deferred` + `runAsync` with a side-effecting callback?)(@SystemFw )

Reply:
This makes sense to me, see comment below about `shift`. Also in CE3 we might want to do this differently, but it's too hard to think about both at this stage, let's get this going instead

Reply:
Cool, so are you happy for me to go ahead and merge this?

Reply:
yeah I would